core/consensus/grandpa/impl/grandpa_impl.cpp

```diff
@@ -477,10 +594,52 @@ namespace kagome::consensus::grandpa {
              msg.round_number,
              peer_id);
     round->doCatchUpResponse(peer_id);
+
+    reputation_repository_->change(peer_id,
+                                   network::reputation::cost::CATCH_UP_REPLY);
   }
 
   void GrandpaImpl::onCatchUpResponse(const libp2p::peer::PeerId &peer_id,
                                       const network::CatchUpResponse &msg) {
+    GrandpaContext::Guard cg;
+
+    auto ctx = GrandpaContext::get().value();
+    if (not ctx->peer_id.has_value()) {
+      if (not pending_catchup_request_.has_value()) {
+        reputation_repository_->change(
+            peer_id, network::reputation::cost::MALFORMED_CATCH_UP);
+        return;
+      }
+
+      const auto &[remote_peer_id, catchup_request] =
+          pending_catchup_request_.value();
+
+      if (peer_id != remote_peer_id) {
+        reputation_repository_->change(
+            peer_id, network::reputation::cost::OUT_OF_SCOPE_MESSAGE);
+        return;
+      }
+
+      if (msg.voter_set_id != catchup_request.voter_set_id
+          or msg.round_number != catchup_request.round_number) {
+        reputation_repository_->change(
+            peer_id, network::reputation::cost::MALFORMED_CATCH_UP);
+        return;
+      }
+
+      if (msg.prevote_justification.empty()
+          or msg.precommit_justification.empty()) {
+        reputation_repository_->change(
+            peer_id, network::reputation::cost::MALFORMED_CATCH_UP);
+        return;
+      }
+
+      auto cleanup = gsl::finally([this] {
+        catchup_request_timer_handle_.cancel();
+        pending_catchup_request_.reset();
+      });
```

Comment:
Already at end of scope
```suggestion
      catchup_request_timer_handle_.cancel();
      pending_catchup_request_.reset();
```

Reply:
Thanks, it is really mistake, but anyway it should be executed when exiting of function.